Tracer can output a compact tracing log file

Dr15Jones opened the pull request:
#### PR description:

Added an option to the Tracer service to output a compact log file of the transitions seen in the framework job. 
Included a python program to convert the file into more humanly readable output.

#### PR validation:

Code compiles. Visual inspection of a file generated from a simple configuration shows the expected values.

Dr15Jones committed:
Added framework transitions to StallMonitor output

Added the option to have the Run and LuminosityBlock transitions added to the file.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42970/37136

- This PR adds an extra 20KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42970/37136/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42970/37136/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42970/37137

- This PR adds an extra 16KB to repository

cmsbuild commented:
A new Pull Request was created by  @Dr15Jones (Chris Jones) for master.

It involves the following packages:

- FWCore/Services (**core**)


@cmsbuild, @makortel, @Dr15Jones, @smuzaffar can you please review it and eventually sign? Thanks.
@wddgit, @makortel, @fwyzard, @missirol this is something you requested to watch as well.
@rappoccio, @antoniovilela, @sextonkennedy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

Dr15Jones commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-3882c9/35110/summary.html
**COMMIT**: 0ec91093118b67682bdfbab5e227d1fbe5b5fa53
**CMSSW**: CMSSW_13_3_X_2023-10-09-2300/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/42970/35110/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially removed 3 lines from the logs
 - Reco comparison results: 8 differences found in the comparisons
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 3356920
 - DQMHistoTests: Total failures: 3
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3356895
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 214 log files, 167 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_3_X_2023-10-09-2300+3882c9/59180/triggerResults): no differences found

Dr15Jones commented:
hold

considering putting this ability elsewhere

cmsbuild commented:
Pull request has been put on hold by  @Dr15Jones
They need to issue an `unhold` command to remove the `hold` state or L1 can `unhold` it for all

Dr15Jones committed:
Moved StallMonitor code to monitor_file_utilities

This will allow StallMonitor and Tracer to share code for creating monitor files.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42970/37167

- This PR adds an extra 20KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42970/37167/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42970/37167/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42970/37168

- This PR adds an extra 20KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42970/37168/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42970/37168/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

Dr15Jones commented:
code-checks

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42970/37169

- This PR adds an extra 20KB to repository

cmsbuild commented:
Pull request #42970 was updated. @smuzaffar, @Dr15Jones, @cmsbuild, @makortel can you please check and sign again.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42970/37186

- This PR adds an extra 20KB to repository

cmsbuild commented:
Pull request #42970 was updated. @Dr15Jones, @makortel, @smuzaffar, @cmsbuild can you please check and sign again.

Dr15Jones committed:
Allow Services to see the list of EventSetup Records

Added the activity eventSetupConfiguration to ActivityRegistry.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42970/37224

- This PR adds an extra 60KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42970/37224/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42970/37224/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

cmsbuild commented:
Pull request #42970 was updated. @smuzaffar, @cmsbuild, @makortel, @Dr15Jones can you please check and sign again.

Dr15Jones committed:
Tracer service can write a compact tracer file

The file contains all the tracer state information with time stamps

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42970/37275

- This PR adds an extra 84KB to repository

cmsbuild commented:
Pull request #42970 was updated. @makortel, @cmsbuild, @smuzaffar, @Dr15Jones can you please check and sign again.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42970/37283

- This PR adds an extra 88KB to repository

cmsbuild commented:
Pull request #42970 was updated. @smuzaffar, @makortel, @Dr15Jones, @cmsbuild can you please check and sign again.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42970/37305

- This PR adds an extra 32KB to repository

cmsbuild commented:
Pull request #42970 was updated. @smuzaffar, @makortel, @cmsbuild, @Dr15Jones can you please check and sign again.

Dr15Jones committed:
Properly set Context when prefetching ES data

When fetching an EventSetup data product during a stream Run
or LuminosityBlock transition, the proper ModuleCallingContext
needs to be set.

Dr15Jones commented:
please test

Dr15Jones commented:
untold

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42970/37306

- This PR adds an extra 44KB to repository

cmsbuild commented:
Pull request #42970 was updated. @smuzaffar, @Dr15Jones, @makortel can you please check and sign again.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-3882c9/35321/summary.html
**COMMIT**: 2e2966893785139cbc97ff43b0532ac337bb7166
**CMSSW**: CMSSW_13_3_X_2023-10-20-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/42970/35321/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially removed 5 lines from the logs
 - Reco comparison results: 30 differences found in the comparisons
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 3357400
 - DQMHistoTests: Total failures: 1073
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3356305
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 214 log files, 167 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_3_X_2023-10-20-1100+3882c9/59343/triggerResults): no differences found

makortel commented on FWCore/Services/bin/edmTracerCompactLogViewer.py:
```diff
@@ -0,0 +1,577 @@
+  Instead of using the StallMonitor Service, you can use the
+  Tracer Service.  Make sure to use the 'printTimestamps' option
+  cms.Service("Tracer", printTimestamps = cms.untracked.bool(True))
+  There are problems associated with this and it is not recommended.'''
```
How accurate is this help message? By quick look it looks mostly being copied from `edmStreamStallGrapher.py`.

Dr15Jones replied:
fixed

makortel commented on FWCore/Services/bin/edmTracerCompactLogViewer.py:
```diff
@@ -0,0 +1,577 @@
+#  postESModuleAcquire = 'b',
+#  preFrameworkTransition = 'F',
+#  postFrameworkTransition = 'f'
+#};
```
Is this commented out piece useful here?

makortel replied:
Would it be useful define these as "enum" similarly to the `Phase` class?

Dr15Jones replied:
I hadn't done that as the values are used in one and only one spot in the entire code.

makortel commented on FWCore/Services/bin/edmTracerCompactLogViewer.py:
```diff
@@ -0,0 +1,577 @@
+kSourceFindEvent = "sourceFindEvent"
+kSourceDelayedRead ="sourceDelayedRead"
+
+class Phase (object):
```
Maybe add a comment that the values must correspond to those of `enum class Phase` in `FWCore/Services/plugins/tracer_setupFile.cc`?

Dr15Jones replied:
done

makortel commented on FWCore/Services/plugins/tracer_setupFile.cc:
```diff
@@ -0,0 +1,814 @@
+  using clock_t = std::chrono::steady_clock;
+  auto now = clock_t::now;
+
+  enum class step : char {
```
Could this enum start with upper case letter, i.e. `Step`?

Dr15Jones replied:
done

makortel commented on FWCore/Services/plugins/tracer_setupFile.cc:
```diff
@@ -0,0 +1,814 @@
+namespace {
+  using duration_t = std::chrono::microseconds;
+  using clock_t = std::chrono::steady_clock;
+  auto now = clock_t::now;
```
Can this global variable be `const`? Or use a function for the shorthand?

Dr15Jones replied:
done

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42970/37372

- This PR adds an extra 24KB to repository

cmsbuild commented:
Pull request #42970 was updated. @smuzaffar, @cmsbuild, @makortel, @Dr15Jones can you please check and sign again.

Dr15Jones commented:
please test

makortel commented:
The unit test fails in IBs

makortel commented:
+core

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (but tests are reportedly failing). This pull request will now be reviewed by the release team before it's merged. @rappoccio, @antoniovilela, @sextonkennedy (and backports should be raised in the release meeting by the corresponding L2)

Dr15Jones commented:
please test

Dr15Jones commented:
NOTE: It is impossible for this PR to affect any test results as the change can only be seen if one uses a brand new parameter in the Tracer service, of which no code in CMSSW uses.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-3882c9/35481/summary.html
**COMMIT**: a081e6f6fcc8f46e22c905b8281f5d711d8a0ef4
**CMSSW**: CMSSW_13_3_X_2023-10-28-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/42970/35481/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially removed 11 lines from the logs
 - Reco comparison results: 30 differences found in the comparisons
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 3362691
 - DQMHistoTests: Total failures: 1073
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3361596
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 214 log files, 167 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_3_X_2023-10-28-1100+3882c9/59473/triggerResults): no differences found

antoniovilela commented:
+1

makortel commented:
Comparison differences are related to https://github.com/cms-sw/cmssw/issues/39803
